### ISSUE TITLE
Ragel 7.0.0.6 can't resolve certain state references

### ISSUE THREAD

**Issue by YorickPeterse** [colmnet/ragel#4]
Consider the following machine:

```ragel
%%machine test;

%% write data;

pub fn dummy() {
    %% write init;
    %% write exec;
}

%%{
    foo := |*
        any => { fnext bar; };
    *|;

    bar := |*
        any => { fnext main; };
    *|;

    main := |*
        any => { fnext foo; };
    *|;
}%%
```

Compiling this with `ragel -F1 -U /tmp/test.rl -o /tmp/test.rs` results in the following output:

```
/tmp/test.rl:12:18: could not resolve state reference 0x26aec40
/tmp/test.rl:20:18: could not resolve state reference 0x26afbb0
```

As far as I can tell this _should_ work fine, though perhaps I'm wrong.

**Comment by YorickPeterse** [comment]
If I instead use this:

```
%%machine test;

%% write data;

pub fn dummy() {
    %% write init;
    %% write exec;
}

%%{
    bar := |*
        any => { fnext main; };
    *|;

    foo := |*
        any => { fnext bar; };
    *|;

    main := |*
        any => { fnext foo; };
    *|;
}%%
```

The output is now:

    /tmp/test.rl:16:18: could not resolve state reference 0x286d3d0

**Comment by YorickPeterse** [comment]
This problem persists when building from Git (as of commit 4ac084dbc8717c0e7051be8b436206438266b0c4) instead.

**Comment by colmnet** [comment]
I can reproduce the printing of a pointer where it should be a string name ref, but not the failed lookups. Can you verify you are working with the latest colm? Next question would be about your platform.

**Comment by YorickPeterse** [comment]
```
$ ragel --version
Ragel State Machine Compiler version 7.0.0.6 January 2016
Copyright (c) 2001-2015 by Adrian Thurston
$ colm --version
Colm version 0.13.0.3 Jan 2016
Copyright (c) 2007-2012 by Adrian D. Thurston
```

Ragel is built from Git as of commit 16365694b74a444ab13c78bc291a3592ed51cb70. The command I ran to test this is `ragel -C -G0 /tmp/test.rl -o /tmp/test.rs` where `/tmp/test.rl` points to the above sample snippet. My platform is Arch Linux, uname output:

    Linux x1carbon 4.4.1-2-ARCH #1 SMP PREEMPT Wed Feb 3 13:12:33 UTC 2016 x86_64 GNU/Linux

**Comment by colmnet** [comment]
best guess right now is that correct compilation is dependent on a sort of memory addresses, somewhere in the new code

**Comment by colmnet** [comment]
I'm still unable to reproduce this. Tried with various optimization combinations between ragel and colm.

What g++ version are you using?

**Comment by YorickPeterse** [comment]
My g++ version is as following:

```
g++ (GCC) 5.3.0
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by YorickPeterse** [comment]
It's worth mentioning that the above bug _only_ occurs when there are at least 3 machines (e.g. foo, bar, and main), it _does_ work when there are 2 or less.

**Comment by colmnet** [comment]
Just installed gcc 5.3 and I'm still unable to reproduce this, so that's not the issue.

**Comment by colmnet** [comment]
Problem was newer C++ stdlib. Reproduced and fixed in 9be749f2fb04932c0c6d49e81651bcf0c6be5fec

**Comment by YorickPeterse** [comment]
@colmnet Sweet, thanks!
